File: ios/App/App/ExtendedFirebasePlugin.swift
//
//  FirebaseAppCheckPlugin.swift
//  App
//
//  Created by Jarret on 20/03/2026.
//

import Foundation
import Capacitor
import FirebaseAppCheck

@objc(ExtendedFirebasePlugin)
public class ExtendedFirebasePlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "ExtendedFirebase"
    public let jsName = "ExtendedFirebase"

    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getAppCheckToken", returnType: CAPPluginReturnPromise),
    ]

    @objc func getAppCheckToken(_ call: CAPPluginCall) {
        let forceRefresh = call.getBool("forceRefresh") ?? false

        FirebaseAppCheck.shared.getToken(forceRefresh: forceRefresh) { result in
            switch result {
            case let .success(token):
                var result = JSObject()
                result["token"] = token
                call.resolve(result)

            case let .failure(error):
                call.resolve()
            }
        }
    }
}

File: android/app/src/main/java/com/whoisup/app/ExtendedFirebasePlugin.kt
package com.whoisup.app

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

@CapacitorPlugin(name = "ExtendedFirebase")
class ExtendedFirebasePlugin : Plugin() {
    @PluginMethod
    fun getAppCheckToken(call: PluginCall) {
        val forceRefresh = call.getBoolean("forceRefresh", false) ?: false

        CoroutineScope(Dispatchers.IO).launch {
            val token = FirebaseAppCheck.getToken(forceRefresh)
            val result = JSObject()
            result.put("token", token)
            call.resolve(result)
        }
    }
}